src/libinfinistore.cpp

```diff
@@ -719,39 +646,49 @@ int rw_rdma(connection_t *conn, char op, std::vector<block_t> &blocks, int block
     conn->cv.wait(lock,
                   [&conn, &blocks] { return conn->rdma_inflight_count + blocks.size() <= MAX_WR; });
 
-    for (int i = 0; i < response.blocks.size(); i++) {
-        DEBUG("remote response: addr: {}, rkey: {}", response.blocks[i].remote_addr,
-              response.blocks[i].rkey);
-
-        // request_mr could be temperary mr or registered mr
+    int cnt = response.blocks.size();
+    struct ibv_send_wr wr[cnt], *bad_wr = nullptr;
+    struct ibv_sge sge[cnt];
+    for (int i = 0; i < cnt; i++) {
         IBVMemoryRegion *request_mr = NULL;
         if (conn->limited_bar1) {
             request_mr = search_mr_from_ptr(local_mr, (char *)base_ptr + blocks[i].offset);
         }
         else {
             request_mr = search_mr_from_ptr(conn->local_mr, (char *)base_ptr + blocks[i].offset);
         }
-        int ret;
-        if (op == OP_RDMA_WRITE) {
-            ret = perform_rdma_write(conn, (char *)base_ptr + blocks[i].offset, block_size,
-                                     response.blocks[i].remote_addr, block_size,
-                                     response.blocks[i].rkey, request_mr);
-        }
-        else if (op == OP_RDMA_READ) {
-            ret = perform_rdma_read(conn, response.blocks[i].remote_addr, block_size,
-                                    (char *)base_ptr + blocks[i].offset, block_size,
-                                    response.blocks[i].rkey, request_mr);
+
+        sge[i].addr = (uintptr_t)((char *)base_ptr + blocks[i].offset);
+        sge[i].length = block_size;
+        sge[i].lkey = request_mr->get_mr()->lkey;
+
+        if (conn->limited_bar1) {
+            // we have to pass mr to cq_handler to deregister it
+            wr[i].wr_id = (uint64_t)request_mr;
+            request_mr->add_ref();
         }
         else {
-            ERROR("Invalid operation");
-            return -1;
-        }
-        if (ret < 0) {
-            ERROR("Failed to perform RDMA operation");
-            return -1;
+            // maybe we will use request ctx in the future
+            wr[i].wr_id = (uintptr_t)conn;
         }
+
+        wr[i].opcode = op == OP_RDMA_READ ? IBV_WR_RDMA_READ : IBV_WR_RDMA_WRITE;
+        wr[i].sg_list = &sge[i];
+        wr[i].num_sge = 1;
+        wr[i].send_flags = IBV_SEND_SIGNALED;
```

Comment:
it should be signaled every a few ops.  Not everyone

Reply:
any document refer to this? 
each WR should generate a WC, right?

Reply:
from chatgpt. this is how we could reduce number of CQ events.